### PR TITLE
Add popup for public room promotions and demotions

### PR DESCRIPTION
**Pull request by Hidden50**
Sadly, no desktop notification will be created by user.popup().
Might add that, too, if I find a good way to do it.

### CODE REVIEW

**Comment by Hidden50** [comment]
still missing a space in one of the messages - I'll amend it in a bit.

**Comment by JohnS66139** [comment]
you need to rebase(there's a guide on the roomintro of the room)

**Comment by Hidden50** [comment]
Do I? Looks like just updating it was fine?

**Comment by JohnS66139** [comment]
so it's not 4 commits for what could easily fit in 1

File: commands.js [hunk 1]
```diff
@@ -722,6 +722,7 @@ exports.commands = {
 
 		room.auth[targetUser.userid] = '#';
 		this.addModCommand("" + name + " was appointed Room Owner by " + user.name + ".");
+		targetUser.popup("You were appointed Room Owner by " + user.name + " in " + room + ".");
```
Comment (sirDonovan) [inline commands.js]: <code>room</code> is an object in these commands so you should use <code>room.id</code> instead.

File: commands.js [hunk 2]
```diff
@@ -744,6 +745,7 @@ exports.commands = {
 
 		delete room.auth[userid];
 		this.sendReply("(" + name + " is no longer Room Owner.)");
+		targetUser.popup("You are no longer Room Owner of " + room + ". (Demoted by " + user.name + ".)");
```
Comment (sirDonovan) [inline commands.js]: This would crash if <code>targetUser</code> did not exist (see relevant line below).

File: commands.js [hunk 3]
```diff
@@ -818,8 +820,10 @@ exports.commands = {
 			if (targetUser && Rooms.rooms[room.id].users[targetUser.userid]) targetUser.popup("You were demoted to Room " + groupName + " by " + user.name + ".");
 		} else if (nextGroup === '#') {
 			this.addModCommand("" + name + " was promoted to " + groupName + " by " + user.name + ".");
+			targetUser.popup("You were promoted to " + groupName + " by " + user.name + ".");
```
Comment (sirDonovan) [inline commands.js]: Same crash as above.

File: commands.js [hunk 4]
```diff
@@ -818,8 +820,10 @@ exports.commands = {
...
 		} else {
 			this.addModCommand("" + name + " was promoted to Room " + groupName + " by " + user.name + ".");
+			targetUser.popup("You were promoted to Room " + groupName + " by " + user.name + " in " + room + ".");
```
Comment (sirDonovan) [inline commands.js]: You should check that the user is in the room to avoid issues with <code>/roomvoice</code>. Also same crash here as above.

**Comment by kotarou3** [comment]
I think you meant fixup

**Comment by Zarel** [comment]
Please do not create new pull requests; just force-push to the pull request you already have open. Creating new pull requests spams all 156 subscribers to this repository and is completely unnecessary.

**Comment by Zarel** [comment]
This is also a bad idea in private rooms and groupchats, where people could promote/demote you to spam you.

**Comment by Zarel** [comment]
Ahem:

> This is also a bad idea in private rooms and groupchats, where people could promote/demote you to spam you.

**Comment by kotarou3** [comment]
Make it such that only if the user is in the room, will you get a popup

**Comment by Zarel** [comment]
I'm fine with the above change.

**Comment by Hidden50** [comment]
If all concerns (see below) have been addressed, this should be ready to merge(?)

--

>Zarel commented 16 hours ago
This is also a bad idea in private rooms and groupchats, where people could promote/demote you to spam you.

 Changed it to work in public, non-personal, non-battle rooms only. (That's what I meant to do anyway, but when I got to it yesterday, I forgot.)
Note that this was indeed a problem for demotions in private rooms, which would produce popups before and could be used to spam.

--

@sirDonovan: room.id and testing for (!targetUser) have been added.

File: commands.js [hunk 5]
```diff
@@ -744,6 +745,7 @@ exports.commands = {
 
 		delete room.auth[userid];
 		this.sendReply("(" + name + " is no longer Room Owner.)");
+		if (targetUser) targetUser.popup("You are no longer a Room Owner of " + room.id + ". (Demoted by " + user.name + ".)");
 		if (targetUser) targetUser.updateIdentity();
```
Comment (Zarel) [inline commands.js]: Should be combined with above line.

File: commands.js [hunk 6]
```diff
@@ -813,13 +815,22 @@ exports.commands = {
 			room.auth[userid] = nextGroup;
 		}
 
+		// Only show popup if: user is online and in the room, the room is public, and not a groupchat or a battle.
+		let needsPopup = targetUser && Rooms.rooms[room.id].users[targetUser.userid] && !room.isPrivate && !room.isPersonal && !room.battle;
```
Comment (Zarel) [inline commands.js]: `Rooms.rooms[room.id].users` can be replaced with `room.users`.

@TheImmortal had a brainfart when he wrote that code, but we might as well fix it while we're refactoring here.

**Comment by Zarel** [comment]
Other than those two, this looks good!

**Comment by Zarel** [comment]
Can I get an MIT license from you? Repeat after me:

> I, ____, hereby license my contributions to the Pokémon Showdown project, past and future, under the MIT license, available in the LICENSE file in this repository.

**Comment by Hidden50** [comment]
I, Daniel Kilian, hereby license my contributions to the Pokémon Showdown project, past and future, under the MIT license, available in the LICENSE file in this repository.

**Pull request merged**